src/Database/LSMTree/Internal.hs

```diff
@@ -235,16 +267,33 @@ openSession hfs hbio dir = do
 
 {-# SPECIALISE closeSession :: Session IO HandleIO -> IO () #-}
 -- | See 'Database.LSMTree.Common.closeSession'.
+--
+-- A session's global resources will only be released once it is sure that no
+-- tables are open anymore.
 closeSession ::
      m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
   => Session m h
   -> m ()
 closeSession Session{sessionState} =
     RW.modify_ sessionState $ \case
       SessionClosed -> pure SessionClosed
-      SessionOpen SessionEnv {sessionLockFile} -> do
-        -- TODO: close known table handles
-        System.hClose sessionLockFile
+      SessionOpen seshEnv -> do
+        -- Since we have a write lock on the session state, we know that no
+        -- tables will be added while we are closing the session, and that we
+        -- are the only thread currently closing the session.
+        --
+        -- We technically don't have to overwrite this with an empty Map, but
+        -- why not
+        tables <- modifyMVar (sessionOpenTables seshEnv) (\m -> pure (Map.empty, m))
```

Comment:
Doesn't matter much either way. As long as reading from the `MVar` is protected by the RW lock, no-one should be able to observe the difference.